src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor.go

```diff
@@ -5,320 +5,315 @@ import (
-		amountToSlash = currEntry.Impl().LockedPledgeCollateral_
-		currEntry.Impl().LockedPledgeCollateral_ = 0
-		// TODO: extra handling of not having enough pledgecollateral to be slashed
-	} else {
-		currEntry.Impl().LockedPledgeCollateral_ = currEntry.LockedPledgeCollateral() - amount
-	}
+	// TODO: call constructor of StorageMinerActor
+	// store ownerAddr and workerAddr there
+	// and return StorageMinerActor address
 
-	// TODO: send amountToSlash to TreasuryActor
-	panic(amountToSlash)
-	spa.PowerTable_[minerID] = currEntry
+	// TODO: minerID should be a MinerActorID
+	// which is smaller than MinerAddress
+	var minerID addr.Address
+	st.PowerTable()[minerID] = newMiner
+
+	h.Commit(st)
+
+	return minerID
 
-	// TODO: commit state change
 }
 
-// TODO: batch process this if possible
-func (spa *StoragePowerActor_I) lockPledgeCollateral(address addr.Address, amount actor.TokenAmount) {
-	// AvailableBalance -> LockedPledgeCollateral
-	// TODO: potentially unnecessary check
-	if amount < 0 {
-		// TODO: proper throw
-		panic("TODO")
-	}
+func (a *StoragePowerActorCode_I) RemoveStorageMiner(rt Runtime, address addr.Address) {
 
-	// TODO: convert address to MinerActorID
+	h, st := a.State(rt)
+
+	// TODO: make explicit address type
 	var minerID addr.Address
 
-	currEntry, found := spa.PowerTable()[minerID]
-	if !found {
-		// TODO: proper throw
-		panic("TODO")
+	if (st.PowerTable()[minerID].ActivePower() + st.PowerTable()[minerID].InactivePower()) > 0 {
+		rt.Fatal("spa.RemoveStorageMiner: power still remains.")
 	}
 
-	if currEntry.Impl().AvailableBalance() < amount {
-		// TODO: proper throw cannot lock more than one has available
-		panic("TODO")
+	delete(st.PowerTable(), minerID)
+
+	h.Commit(st)
+}
+
+func (a *StoragePowerActorCode_I) GetTotalPower(rt Runtime) block.StoragePower {
+
+	h, st := a.State(rt)
+
+	totalPower := block.StoragePower(0)
+	for _, miner := range st.PowerTable() {
+		totalPower = totalPower + miner.ActivePower() + miner.InactivePower()
 	}
 
-	currEntry.Impl().AvailableBalance_ = currEntry.AvailableBalance() - amount
-	currEntry.Impl().LockedPledgeCollateral_ = currEntry.LockedPledgeCollateral() + amount
-	spa.PowerTable_[minerID] = currEntry
+	h.Release(st)
 
-	// TODO: commit state change
+	return totalPower
 }
 
-func (spa *StoragePowerActor_I) unlockPledgeCollateral(address addr.Address, amount actor.TokenAmount) {
-	// lockedPledgeCollateral -> AvailableBalance
-	if amount < 0 {
-		// TODO: proper throw
-		panic("TODO")
-	}
+func (a *StoragePowerActorCode_I) EnsurePledgeCollateralSatisfied(rt Runtime) bool {
 
-	// TODO: convert address to MinerActorID
+	h, st := a.State(rt)
+
+	ret := false
+
+	// TODO: convert msgSender to MinerActorID
 	var minerID addr.Address
 
-	currEntry, found := spa.PowerTable()[minerID]
+	powerEntry, found := st.PowerTable()[minerID]
+
 	if !found {
-		// TODO: proper throw
-		panic("TODO")
+		rt.Fatal("spa.EnsurePledgeCollateralSatisfied: miner not found.")
```

Comment:
@jzimmerman the `h.Commit/Release` structure has one issue -- control flow returns:

```go
h, st := a.State(rt)

err := tryFoo()
if err != nil {
    return rt.ReturnError(err) // did not h.Release or h.Commit
}

h.Commit(st)
```

We should not use `defer` so we should enforce strict locking/unlocking of state, similar to `sync.Mutex`


```go
h, st := a.State(rt)

err := tryFoo()
if err != nil {
    h.Release()
    return rt.ReturnError(err)
}

h.Commit(st)
```

Reply:
- I imagine errors and `rt.Fatal` _can_ automatically release w/o side effect, BUT is automatic undo what we want? or do we want to force spec code writers to make sure they release properly? 
- i lean towards the latter, and calling `rt.ReturnX` to be a panic-worthy error if state has not been released

Reply:
`rt.Fatal` should undo all the state changes and return state to the initial state at state acquisition. Good to make explicit that that is the case but probably dont want to give code writers the option of choosing if state will be released or unrolled?

Reply:
`rt.Fatal` should undo all the state changes and return state to the initial state at state acquisition. Good to make explicit that that is the case but probably dont want to give code writers the option of choosing if state will be released or unrolled?